python-package/xlearn/sklearn.py

```diff
@@ -0,0 +1,185 @@
+    def get_params(self, deep=False):
+        """ Get model parameters
+        :param deep: is deep copy
+        :return: model parameters
+        """
+        params = super(BaseXLearnModel, self).get_params(deep=deep)
+
+        # rename reg_lambda as lambda, and remove model_type
+        params['lambda'] = params.pop('reg_lambda')
+        params.pop('model_type')
+        return params
+
+    def fit(self, X, y):
+        """ Fit the XLearn model given feature matrix X and label y
+        :param X: array-like
+                  Feature matrix
+        :param y: array-like
+                  Label
+        """
+        if self.model_type == 'fm':
+            self._XLearnModel = create_fm()
+        elif self.model_type == 'lr':
+            self._XLearnModel = create_linear()
+        elif self.model_type == 'ffm':
+            self._XLearnModel = create_ffm()
+        else:
+            raise Exception('model_type must be fm, ffm or lr')
+
+        #TODO: find out what task need to set sigmoid
+        if self.task == 'binary':
+            self._XLearnModel.setSigmoid()
+
+        params = self.get_params(deep=True)
+
+        # create temporary files for training data
+        temp_train_file = tempfile.NamedTemporaryFile(delete=True)
+
+        # convert data into libsvm format for training
+        self._convert_data(X, y, temp_train_file.name)
+        self._XLearnModel.setTrain(temp_train_file.name)
+
+        # fit model
+        self._XLearnModel.fit(params, self._temp_model_file.name)
+
+        # remove temporary files for training
+        self._remove_temp_file(temp_train_file)
+
+    def predict(self, X):
+        """ Generate prediction using feature matrix X
+
+        :param X: array-like
+                  Feature matrix
+        :return: prediction
+        """
+        # convert data to libsvm format
+        temp_test_file = tempfile.NamedTemporaryFile(delete=True)
+        self._convert_data(X, None, temp_test_file.name)
+        self._XLearnModel.setTest(temp_test_file.name)
+
+        # generate output
+        temp_output_file = tempfile.NamedTemporaryFile(delete=True)
+        self.get_model().predict(self._temp_model_file.name, temp_output_file.name)
+
+        # read output into numpy
+        pred = np.loadtxt(temp_output_file.name)
+
+        # remove temporary test data and output file
+        self._remove_temp_file(temp_test_file)
+        self._remove_temp_file(temp_output_file)
+
+        return pred
+
+    def feature_importance_(self):
+        """TODO: analyze weight matrix to get feature importance"""
+        pass
+
+    def _convert_data(self, X, y, filepath):
+        """
+        Convert feature matrix X and label y into libsvm format
+        :param X: feature matrix
+        :param y: label or None
+        """
+        if y is None:
+            # create dummy label for test
+            y = np.zeros(X.shape[0], dtype=np.int8)
+
+        try:
+            dump_svmlight_file(X, y, filepath)
```

Comment:
We are writing the in-memory interface for DMatrix class.

Reply:
That would be great. I will keep the internal conversion for now. Once DMatrix class is ready, I can update that accordingly. Is there any expected timeline for the DMatrix class?